### PR TITLE
ECS.ContainerDefinition.StopTimeout docker-compose.service.stop_grace_period

### LINKED ISSUE #717: [FR] ECS.ContainerDefinition.StopTimeout docker-compose.service.stop_grace_period
#### Feature Request

Some containers (e.g. postgres database amongst others) need a longer time before forceful shutdown in order to cleanly save data.

The docker-compose.yml service `stop_grace_period` field should be translated to ECS Container Definition `StopTimeout`.

#### Documentation Links
- https://docs.docker.com/compose/compose-file/compose-file-v3/#stop_grace_period
- https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinition.html#cfn-ecs-taskdefinition-containerdefinition-stoptimeout

### PR DESCRIPTION
**Pull request by thorfi**
Set ECS.ContainerDefinition.StopTimeout from docker-compose.service.stop_grace_period 

Fixes #717

### CODE REVIEW

**Comment by thorfi** [comment]
Hi @JohnPreston ,

I tried to stick to naming/logging conventions as far as I understand them, but let me know if anything needs tweaking before you merge.

Thanks!

File: ecs_composex/compose/compose_services/__init__.py [hunk 1]
```diff
@@ -754,6 +754,25 @@ def ulimits(self) -> Union[list, Ref]:
...
+    def stop_grace_period(self):
+        """
+        Method to import and determine StopTimeout
+        """
+        __stop_grace_period = set_else_none("stop_grace_period", self.definition)
+        if not __stop_grace_period:
+            return NoValue
+        if not isinstance(__stop_grace_period, (int, str, float)):
+            raise TypeError(self.name)
+        stop_timeout = import_time_values_to_seconds(__stop_grace_period)
```
Comment (JohnPreston) [inline ecs_composex/compose/compose_services/__init__.py:767]:
```suggestion
        stop_timeout = import_time_values_to_seconds(__stop_grace_period, 120)
```
the function `import_time_values_to_seconds` already handles to set a max value. Given ECS max value is 120, I'd set it there? Or did you handle it below to warn the user?

File: ecs_composex/compose/compose_services/__init__.py [hunk 2]
```diff
@@ -754,6 +754,25 @@ def ulimits(self) -> Union[list, Ref]:
...
 
+    @property
+    def stop_grace_period(self):
+        """
+        Method to import and determine StopTimeout
+        """
+        __stop_grace_period = set_else_none("stop_grace_period", self.definition)
+        if not __stop_grace_period:
+            return NoValue
+        if not isinstance(__stop_grace_period, (int, str, float)):
```
Comment (JohnPreston) [inline ecs_composex/compose/compose_services/__init__.py:765]: According to https://github.com/compose-spec/compose-spec/blob/master/schema/compose-spec.json#L356 the type has to be a string.
Given compose-x uses the spec to do validation, we probably will never get any other value than a string, but certainly shouldn't.
```suggestion
        if not isinstance(__stop_grace_period, (str)):
```

**Review by JohnPreston — COMMENTED** [review]
Hi @thorfi 
Great stuff as always, thanks for the contribution.
See my notes below and let me know

File: ecs_composex/compose/compose_services/__init__.py (hunk 1 above)
Comment (thorfi) [reply ecs_composex/compose/compose_services/__init__.py:767]: I did handle it below so in order to warn the user of too short/long, but I'm happy to do it without the warning if you prefer?

**Comment by JohnPreston** [comment]
@thorfi can you please fix the indentation/formatting with black?
For future ease you your end, use pre-commit

```bash
python3 -m pip install pip -U; python -m pip install pre-commit
cd ecs_composex; pre-commit install
```
That way pre-commit will trigger when you do `git commit` and auto correct anything it can.

**Comment by thorfi** [comment]
> @thorfi can you please fix the indentation/formatting with black? For future ease you your end, use pre-commit

Done! Thanks

**Pull request merged**